src/vs/base/common/arrays.ts

```diff
@@ -589,15 +589,119 @@ export function maxIndex<T>(array: readonly T[], fn: (value: T) => number): numb
 	return maxIdx;
 }
 
-/**
- * Returns an array of all elements `e` with `startKey <= keySelector(e) <= endKey`.
- * `data` must be sorted in ascending order.
-*/
-export function filterSortedByKey<T>(data: readonly T[], keySelector: (value: T) => number, startKey: number, endKey = startKey): T[] | null {
-	const startIdx = findFirstInSorted(data, x => startKey <= keySelector(x));
-	const endIdx = findFirstInSorted(data, x => endKey < keySelector(x));
-	if (startIdx === endIdx) {
-		return null;
-	}
-	return data.slice(startIdx, endIdx);
+export class ArrayQueue<T> {
+	private firstIdx = 0;
+	private lastIdx = this.items.length - 1;
+
+	/**
+	 * Constructs a queue that is backed by the given array. Runtime is O(1).
+	*/
+	constructor(private readonly items: T[]) { }
+
+	get length(): number {
+		return this.lastIdx - this.firstIdx + 1;
+	}
+
+	/**
+	 * Consumes elements from the beginning of the queue as long as the predicate returns true.
+	 * If no elements were consumed, `null` is returned. Has a runtime of O(result.length).
+	 *
+	 * If isPredicateMonotonous is set to `true` and predicate is monotonous on the data,
+	 * it has a runtime of O(log(result.length)) + time(slice(result.length)).
+	*/
+	takeWhile(predicate: (value: T) => boolean, isPredicateMonotonous: boolean = false): T[] | null {
+		// P(k) := k <= this.lastIdx && predicate(this.items[k])
+		// Find s := min { k | k >= this.firstIdx && !P(k) } and return this.data[this.firstIdx...s)
+
+		if (!isPredicateMonotonous) {
+			let startIdx = this.firstIdx;
+			while (startIdx < this.items.length && predicate(this.items[startIdx])) {
+				startIdx++;
+			}
+			const result = startIdx === this.firstIdx ? null : this.items.slice(this.firstIdx, startIdx);
+			this.firstIdx = startIdx;
+			return result;
+		} else {
+			// `predicate` is monotonous, i.e. `!P(k) => !P(k + 1)`
+
+			let startIdx = this.firstIdx;
+			let delta = 1;
+			while (startIdx + delta <= this.lastIdx && predicate(this.items[startIdx + delta])) {
+				delta = delta * 2;
```

Comment:
This is a very nice idea. I'm not sure how to use big `O` notation to express this one though :). But one caveat is that despite being faster on paper, a linear loop here might actually be faster here for small number of elements (I think due to random memory access patterns).

I recommend looking at recent presentations from Andrei Alexandrescu (e.g. https://www.youtube.com/watch?v=FJJTYQYB1JQ ) where he brings up the limitations of the big `O` notation in general and improves the runtime of `std::sort` by doing more work than necessary :).

Reply:
> This is a very nice idea. I'm not sure how to use big O notation to express this one though :).
It is `O(log(result.length))` 😉 - the runtime complexity depends on the output size and is not easily derivable from the input directly (but the output is independent of the algorithm, in particular its complexity).
`delta` is doubled at most `log2(result.length) + 1` times, with `result.length` being the length of the returned array.

>  linear loop here might actually be faster here for small number of elements (I think due to random memory access patterns).

Not neccesarily. For small number of elements, all the element pointers should fit into the cache anyway. For large number of elements, you need to refetch the cache anyway.
But I guess this is best verified empirically 😉

Reply:
I suggest that we write some benchmarks and measure runtimes.

I recommend https://www.youtube.com/watch?v=FJJTYQYB1JQ . One of my key takeaways from that video is that assumptions might not hold true, even if they make sense, and measuring is far superior to any kind of pen & paper analysis.

Reply:
You are right 😢 It is a little bit slower for small inputs, though I don't understand why.
It is significantly faster though on large inputs.

Anyway, I suggest to remove the exponential search (and keep it around somewhere). It was a nice exploration.